client/feature-flags.json

```diff
@@ -0,0 +1,6 @@
+[
+  {
```

Comment:
Why not store as key-value map? (Also, any value in allowing feature flags to be modified through URL params? This has helped us test in production before removing feature flag and exposing by default.)

Reply:
I was thinking about this structure, however if we were to move this logic to backend and have a graphql endpoint, you would have to change the type every time a new feature flag is added. The way its implemented at the moment allows typing that doesn't change and can be extended without having to worry about updating schemas

Reply:
Also URL params I think can be a bit dangerous if left in for production. My future plan will be having per environment feature flag configs that can be swapped out at deploy time.

Reply:
What would be the benefit of returning this from graphql endpoint? (graphql supports JSON scalar, but it's an all or nothing query)

Are you worried about your users misusing feature flags? It sounds "dangerous", but actually in practice can be easy to use and allow you to verify your code in different environments pretty easy.

Reply:
I don't think feature flags should be controlled from the client.

This opens up potential for misuse of feature flags. As authorization is currently not supported it can be dangerous in production because it allows actions on workflows which users may not have the right to do.

Authentication and authorization is coming soon which should alleviate some of these concerns, but is not currently live in production yet.

Reply:
> it allows actions on workflows which users may not have the right to do

I would suggest service-level authorization checks to make sure users have the right to use actions, rather than relying on the UI to manage that through feature flags. We use feature flags in a similar way to allow new actions but the service still validates that the user has authorization to add that new action when it's used. It doesn't prevent the user from seeing the action or trying to add it, but the service blocks them when they do.

Perhaps we should consider what types of feature flags are needed - semi-permanent flags that require deployment, or test toggles that can be activated or deactivated in the browser to test a new feature?

The feature-flag helper can offer all use cases (deployment flags, environment flags, URL flags, and other types).

Reply:
Agreed. Once authorization is implemented on server side, a lot of this code can be made more flexible. Thanks.